src/vos/vos_common.c

```diff
@@ -315,102 +320,76 @@ vos_tx_end(struct vos_container *cont, struct dtx_handle *dth_in,
+static void
+vos_tls_fini(const struct dss_thread_local_storage *dtls,
+	     struct dss_module_key *key, void *data)
 {
-	if (imem_inst->vis_ocache)
-		vos_obj_cache_destroy(imem_inst->vis_ocache);
+	struct vos_tls *tls = data;
+
+	D_ASSERT(d_list_empty(&tls->vtl_gc_pools));
+	if (tls->vtl_ocache)
+		vos_obj_cache_destroy(tls->vtl_ocache);
 
-	if (imem_inst->vis_pool_hhash)
-		d_uhash_destroy(imem_inst->vis_pool_hhash);
+	if (tls->vtl_pool_hhash)
+		d_uhash_destroy(tls->vtl_pool_hhash);
 
-	if (imem_inst->vis_cont_hhash)
-		d_uhash_destroy(imem_inst->vis_cont_hhash);
+	if (tls->vtl_cont_hhash)
+		d_uhash_destroy(tls->vtl_cont_hhash);
+
+	umem_fini_txd(&tls->vtl_txd);
+	D_FREE(tls);
 }
 
-static inline int
-vos_imem_strts_create(struct vos_imem_strts *imem_inst)
+static void *
+vos_tls_init(const struct dss_thread_local_storage *dtls,
+	     struct dss_module_key *key)
 {
-	int		rc;
+	struct vos_tls *tls;
+	int rc;
+
+	D_ALLOC_PTR(tls);
+	if (tls == NULL)
+		return NULL;
 
-	rc = vos_obj_cache_create(LRU_CACHE_BITS,
-				  &imem_inst->vis_ocache);
+	D_INIT_LIST_HEAD(&tls->vtl_gc_pools);
+	rc = vos_obj_cache_create(LRU_CACHE_BITS, &tls->vtl_ocache);
 	if (rc) {
 		D_ERROR("Error in creating object cache\n");
-		return rc;
+		goto failed;
 	}
 
 	rc = d_uhash_create(D_HASH_FT_NOLOCK, VOS_POOL_HHASH_BITS,
-			    &imem_inst->vis_pool_hhash);
+			    &tls->vtl_pool_hhash);
 	if (rc) {
 		D_ERROR("Error in creating POOL ref hash: "DF_RC"\n",
 			DP_RC(rc));
 		goto failed;
 	}
 
 	rc = d_uhash_create(D_HASH_FT_NOLOCK | D_HASH_FT_EPHEMERAL,
-			    VOS_CONT_HHASH_BITS, &imem_inst->vis_cont_hhash);
+			    VOS_CONT_HHASH_BITS, &tls->vtl_cont_hhash);
 	if (rc) {
 		D_ERROR("Error in creating CONT ref hash: "DF_RC"\n",
 			DP_RC(rc));
 		goto failed;
 	}
 
-	return 0;
-
-failed:
-	vos_imem_strts_destroy(imem_inst);
-	return rc;
-}
-
-static void *
-vos_tls_init(const struct dss_thread_local_storage *dtls,
-	     struct dss_module_key *key)
-{
-	struct vos_tls *tls;
-	int rc;
-
-	D_ALLOC_PTR(tls);
-	if (tls == NULL)
-		return NULL;
-
-	D_INIT_LIST_HEAD(&tls->vtl_gc_pools);
-	if (vos_imem_strts_create(&tls->vtl_imems_inst)) {
-		D_FREE(tls);
-		return NULL;
-	}
-
 	rc = umem_init_txd(&tls->vtl_txd);
 	if (rc) {
-		vos_imem_strts_destroy(&tls->vtl_imems_inst);
-		D_FREE(tls);
-		return NULL;
+		D_ERROR("Error in creating txd: %d\n", rc);
+		goto failed;
 	}
 
-	tls->vtl_dth = NULL;
-
 	rc = vos_ts_table_alloc(&tls->vtl_ts_table);
 	if (rc) {
-		umem_fini_txd(&tls->vtl_txd);
-		vos_imem_strts_destroy(&tls->vtl_imems_inst);
-		D_FREE(tls);
-		return NULL;
+		D_ERROR("Error in creating timestamp table: %d\n", rc);
+		goto failed;
 	}
 
 	return tls;
-}
-
-static void
-vos_tls_fini(const struct dss_thread_local_storage *dtls,
-	     struct dss_module_key *key, void *data)
-{
-	struct vos_tls *tls = data;
-
-	vos_imem_strts_destroy(&tls->vtl_imems_inst);
-	umem_fini_txd(&tls->vtl_txd);
-	vos_ts_table_free(&tls->vtl_ts_table);
```

Comment:
The memory leak is due to vos_ts_table_free never being called in new version of vos_tls_fini